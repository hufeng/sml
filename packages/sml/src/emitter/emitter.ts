import path from 'node:path'
import { exec } from 'node:child_process'
import Builder from '../common/builder'

export abstract class Emitter<T> {
  protected sml: T
  protected s: Builder

  constructor(sml: T) {
    this.sml = sml
    this.s = new Builder()
  }

  abstract emitCode(): string

  plantUML(img: string) {
    const jar = path.join(__dirname, '../../bin/plantuml-1.2022.7.jar')

    exec(
      `${this.emitCode()} | java -jar ${jar} -pipe > ${img}`,
      (err, stdout, stderr) => {
        if (err) {
          throw err
        }
        console.log(stdout)
        console.log(stderr)
      },
    )
  }
}
