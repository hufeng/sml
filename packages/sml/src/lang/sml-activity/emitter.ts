import S from '../s'
import { Emitter } from '../base'
import { Component, SmlActivityLangMeta } from './types'

export class PumlActivityEmitter extends Emitter<SmlActivityLangMeta> {
  emitCode() {
    const { components, links, vlinks, notes } = this.meta

    return this.s
      .$reset()
      .$s('@startuml')
      .$fn(this.buildTheme)
      .$fors(components, this.buildComponents)
      .$fors(links, this.buildLinks('->'))
      .$fors(vlinks, this.buildVlink('-->'))
      .$fors(notes, this.buildNotes)
      .$s('@enduml')
      .toString()
  }

  buildComponents(s: S, c: Component) {
    s.$s(`${c.type} "${c.label}" as ${c.id}`)
  }
}
