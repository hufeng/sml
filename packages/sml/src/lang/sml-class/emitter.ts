import { Emitter } from '../base'
import { ClassDiagramAst } from '../types'

export class PumlClassEmitter extends Emitter<ClassDiagramAst> {
  static visibleMap = {
    private: '-',
    protected: '#',
    package_private: '~',
    public: '+',
  } as const

  emitCode() {
    const { clazzes, structs, enums, interfaces, protocols } = this.meta
    this.emitStart()
      .emitInf(interfaces)
      .emitProtocol(protocols)
      .emitClazz(clazzes)
      .emitStruct(structs)
      .emitEnum(enums)
      .emitEnd()

    return this.s.toString('\n')
  }

  private emitStart() {
    this.s.$s('@startuml').$fn(this.buildConfig)
    return this
  }

  private emitEnd() {
    this.s.$s('@enduml')
    return this
  }

  /**
   *
   * @param s
   * @param clazzes
   */
  private emitClazz(clazzes: ClassDiagramAst['clazzes']) {
    this.s.$for(clazzes, (s, clazz) => {
      // extends and implements
      if (clazz.extends.length > 0 || clazz.implements.length > 0) {
        s.$s(
          `class ${clazz.name}${
            clazz.extends.length > 0
              ? ' extends ' + clazz.extends.join(', ')
              : ' '
          }${
            clazz.implements.length > 0
              ? ' implements ' + clazz.implements.join(', ')
              : ' '
          }`,
        )
      }

      // class start
      s.$s(`${clazz.abstract ? 'abstract' : ''} class ${clazz.name} {`)
      // fields
      s.$for(clazz.fields, (s, field) => {
        const visible = PumlClassEmitter.visibleMap[field.visible]
        s.$s(`  ${visible} ${field.name}: ${field.type};`)
      })
      // methods
      s.$for(clazz.methods, (s, method) => {
        const visible = PumlClassEmitter.visibleMap[method.visible]
        s.$s(
          `  ${visible} ${method.abstract ? 'abstract ' : ''}${
            method.name
          }(${method.params.map((p) => `${p.name}: ${p.type}`).join(',')}): ${
            method.ret
          }`,
        )
      })
      s.$s('}')
    })
    return this
  }

  /**
   * emit enums code
   * @param s
   * @param enums
   */
  private emitEnum(enums: ClassDiagramAst['enums']) {
    this.s.$for(enums, (s, e) => {
      s.$s(`enum ${e.name} {`)
        .$for(e.fields, (s, field) =>
          s.$s(
            `  ${field.name}${
              field.value !== 'undefined' ? ': ' + field.value : ''
            }`,
          ),
        )
        .$s('}')
    })

    return this
  }

  private emitInf(infs: ClassDiagramAst['interfaces']) {
    this.s.$for(infs, (s, i) => {
      // implements other interfaces
      if (i.implements.length > 0) {
        s.$s(`interface ${i.name} implements ${i.implements.join(', ')}`)
      }
      // start interface
      s.$s(`interface ${i.name} {`)
        .$for(i.methods, (s, method) => {
          const params = method.params
            .map((p) => `${p.name}: ${p.type}`)
            .join(',')
          s.$s(`  + ${method.name}(${params}): ${method.ret}`)
        })
        .$s('}')
    })
    return this
  }

  private emitProtocol(prots: ClassDiagramAst['protocols']) {
    this.s.$for(prots, (s, prot) => {
      // implements other interfaces
      if (prot.implements.length > 0) {
        s.$s(`interface ${prot.name} implements ${prot.implements.join(', ')}`)
      }
      s.$s(`protocol ${prot.name} {`)
        .$for(prot.methods, (s, method) => {
          const params = method.params
            .map((p) => `${p.name}: ${p.type}`)
            .join(',')
          s.$s(`   + ${method.name}(${params}): ${method.ret}`)
        })
        .$s('}')
    })
    return this
  }

  private emitStruct(structs: ClassDiagramAst['structs']) {
    this.s.$for(structs, (s, struct) => {
      // extends and implements
      if (struct.extends.length > 0 || struct.implements.length > 0) {
        s.$s(
          `class ${struct.name}${
            struct.extends.length > 0
              ? ' extends ' + struct.extends.join(', ')
              : ' '
          }${
            struct.implements.length > 0
              ? ' implements ' + struct.implements.join(', ')
              : ' '
          }`,
        )
      }
      // struct start
      s.$s(`struct ${struct.name} {`)
      // fields
      s.$for(struct.fields, (s, field) => {
        const visible = PumlClassEmitter.visibleMap[field.visible]
        s.$s(`  ${visible} ${field.name}: ${field.type};`)
      })
      // methods
      s.$for(struct.methods, (s, method) => {
        const visible = PumlClassEmitter.visibleMap[method.visible]
        s.$s(
          `  ${visible} ${method.abstract ? 'abstract ' : ''}${
            method.name
          }(${method.params.map((p) => `${p.name}: ${p.type}`).join(',')}): ${
            method.ret
          }`,
        )
      })
      s.$s('}')
    })

    return this
  }
}
