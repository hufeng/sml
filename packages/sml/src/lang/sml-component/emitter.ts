import { Emitter } from '../base'
import { SmlComponentAst } from '../types'

export class PumlComponentEmitter extends Emitter<SmlComponentAst> {
  emitCode() {
    const {
      packages,
      nodes,
      clouds,
      databases,
      components,
      infs,
      links,
      vlinks,
      rels,
    } = this.meta

    return this.s
      .str('@startuml')
      .forStr(
        packages,
        (s, v) =>
          s
            .str(`  package "${v.title}" {`)
            .forStr(v.components, (s, v) =>
              s.str(
                `    component "${v.title}"${v.name ? '  as ' + v.name : ''}`,
              ),
            )
            .forStr(v.infs, (s, v) =>
              s.str(
                `    interface "${v.title}"${v.name ? '  as ' + v.name : ''}`,
              ),
            )
            .str('  }'),
        packages.length > 0 ? '\n' : '',
      )
      .forStr(
        nodes,
        (s, v) =>
          s
            .str(`  node "${v.title}" {`)
            .forStr(v.components, (s, v) =>
              s.str(
                `   component "${v.title}"${v.name ? '  as ' + v.name : ''}`,
              ),
            )
            .forStr(v.infs, (s, v) =>
              s.str(
                `    interface "${v.title}"${v.name ? '  as ' + v.name : ''}`,
              ),
            )
            .str(' }'),
        nodes.length > 0 ? '\n' : '',
      )
      .forStr(
        clouds,
        (s, v) =>
          s
            .str(`  cloud "${v.title}" {`)
            .forStr(v.components, (s, v) =>
              s.str(
                `    component "${v.title}"${v.name ? '  as ' + v.name : ''}`,
              ),
            )
            .forStr(v.infs, (s, v) =>
              s.str(
                `    interface "${v.title}"${v.name ? '  as ' + v.name : ''}`,
              ),
            )
            .str('  }'),
        clouds.length > 0 ? '\n' : '',
      )
      .forStr(
        databases,
        (s, v) =>
          s
            .str(`  database "${v.title}" {`)
            .forStr(v.components, (s, v) =>
              s.str(
                `    component "${v.title}"${v.name ? '  as ' + v.name : ''}`,
              ),
            )
            .forStr(v.infs, (s, v) =>
              s.str(
                `    interface "${v.title}"${v.name ? '  as ' + v.name : ''}`,
              ),
            )
            .str('  }'),
        databases.length > 0 ? '\n' : '',
      )
      .forStr(components, (s, v) =>
        s.str(`  component "${v.title}"${v.name ? '  as ' + v.name : ''}`),
      )
      .forStr(
        infs,
        (s, v) =>
          s.str(`  interface "${v.title}"${v.name ? '  as ' + v.name : ''}`),
        infs.length > 0 ? '\n' : '',
      )
      .forStr(links, (s, v) => s.str(`  ${v.from} --> ${v.to}`))
      .forStr(vlinks, (s, v) => s.str(`  ${v.from} ..> ${v.to}`))
      .forStr(rels, (s, v) => s.str(`  ${v.from} - ${v.to}`))
      .str('@enduml')
      .toString('\n')
  }
}
