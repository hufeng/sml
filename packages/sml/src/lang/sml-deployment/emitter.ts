import S from '../s'
import { Emitter } from '../base'
import { DeploymentLangAst, DeploymentBase, DeploymentContainer } from './types'

export class PumlDeploymentEmitter extends Emitter<DeploymentLangAst> {
  emitCode() {
    const { title, zones, components, links, vlinks, rels } = this.meta

    return this.s
      .reset()
      .$s(`@startuml ${title.replace(/ /g, '_')}`)
      .$fn(this.buildConfig)
      .$s('')
      .$fors(components, this.build)
      .$for(zones, this.buildZones)
      .$for(links, this.buildLinks)
      .$for(vlinks, this.buildVLink)
      .$for(rels, this.buildRels)
      .$s('@enduml')
      .toString()
  }

  build(s: S, v: DeploymentBase) {
    s.$s(`${v.type} "${v.label}"${' as ' + v.id} <<${v.head}>>`)
  }

  buildZones(s: S, v: DeploymentContainer) {
    s.$s(`${v.type} "${v.label}"${' as ' + v.id} <<${v.head}>> {`)
      .$for(v.children, (s, v) =>
        s.$s(`  ${v.type} "${v.label}"${' as ' + v.id} <<${v.head}>>`),
      )
      .$s('}')
  }
}
