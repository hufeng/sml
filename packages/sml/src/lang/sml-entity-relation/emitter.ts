import { Emitter } from '../base'
import S from '../s'
import { Entity, EntityField, EntityRelationDiagramAst, Relation } from './type'

export class PumlEntityRelationEmitter extends Emitter<EntityRelationDiagramAst> {
  constructor(ast: EntityRelationDiagramAst) {
    super(ast)
  }

  emitPuml(): string {
    const { title, entities, relations } = this.meta
    return this.s
      .$reset()
      .$s(`@startuml ${title.replace(/ /g, '_')}`)
      .$s(this.buildTheme)
      .$s(this.buildTile)
      .$s(`hide circle`)
      .$s(`skinparam linetype ortho`)
      .$s('')
      .$for(entities, this.buildEntity)
      .$for(relations, this.buildRelation)
      .$s('@enduml')
      .toString()
  }

  private buildEntity(s: S, e: Entity) {
    return s
      .$s(`entity "${e.label}" as ${e.id} {`)
      .$for(e.ids, (s: S, f: EntityField) =>
        s.$s(`  *${f.name} : ${f.type}`, s.if$(f.stereotypes, ` <<${f.stereotypes}>>`)).$s(`--`),
      )
      .$for(e.fields, (s: S, f: EntityField) =>
        s.$s(`  ${f.name} : ${f.type}`, s.if$(f.stereotypes, ` <<${f.stereotypes}>>`)),
      )
      .$s('}')
  }

  private buildRelation(s: S, rel: Relation) {
    return s.$for(rel.to, (s: S, to) => s.$s(`${rel.from} ${rel.type} ${to}`))
  }
}
