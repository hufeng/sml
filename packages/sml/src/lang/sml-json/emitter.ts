import { Emitter } from '../base'
import { JsonDiagramAst } from './types'

export class PumlJsonEmitter extends Emitter<JsonDiagramAst> {
  emitCode() {
    const { title, highlights, json } = this.meta
    return this.s
      .reset()
      .$s(`@startjson ${title.replace(/ /g, '_')}`)
      .$fn(this.buildTheme)
      .$fors(highlights, (s, highlight) =>
        s.$s(
          `#highlight ${highlight
            .split('.')
            .map((str) => JSON.stringify(str))
            .join(' / ')}`,
        ),
      )
      .$s(JSON.stringify(json, null, 2))
      .$s('@endjson')
      .toString()
  }
}
