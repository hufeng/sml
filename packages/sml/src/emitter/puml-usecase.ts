import { Emitter } from './emitter'
import { SmlUseCase, SmlUseCaseMeta } from '../lang/sml-usecase'

export class PumlUseCaseEmitter extends Emitter<SmlUseCase> {
  emitCode() {
    const meta = (this.sml as any).meta as SmlUseCaseMeta
    return (
      this.s
        // start
        .str('@startuml')
        // setting actor style
        .ifStr(
          meta.actorStyle !== 'default',
          `skinparam actorStyle ${meta.actorStyle}`,
        )
        .str(`${meta.direction.replace('->', ' to ')} direction\n`)
        // actors
        .forStr(
          meta.actors,
          (s, actor) => s.str(`actor :${actor.label}: as ${actor.name}`),
          meta.actors.length > 0 ? '\n' : '',
        )
        //usecases
        .forStr(
          meta.usecases,
          (s, usecase) =>
            s.str(`usecase (${usecase.label}) as ${usecase.name}`).str,
          meta.usecases.length > 0 ? '\n' : '',
        )
        //rects
        .forStr(
          meta.rects,
          (s, r) => {
            s.str(`rectangle ${r.label.replace(/ /g, '_')} {`)
            s.forStr(r.usecases, (s, uc) =>
              s.str(`  usecase "${uc.label}" as ${uc.name}`),
            ).forStr(r.actors, (s, actor) =>
              s.str(`actor :${actor.label}: as ${actor.name}`),
            )
            s.str('}')
          },
          meta.rects.length > 0 ? '\n' : '',
        )
        .forStr(
          meta.pkgs,
          (s, pkg) => {
            s.str(`package ${pkg.label.replace(/ /g, '_')} {`)
            s.forStr(pkg.usecases, (s, uc) =>
              s.str(`  usecase "${uc.label}" as ${uc.name}`),
            ).forStr(pkg.actors, (s, actor) =>
              s.str(`actor :${actor.label}: as ${actor.name}`),
            )
            s.str('}')
          },
          meta.pkgs.length > 0 ? '\n' : '',
        )
        //links
        .forStr(meta.links, (s, link) => {
          s.forStr(link.to, (s, to) => s.str(`${link.from} --> ${to}`))
        })
        //end
        .str('@enduml')
        .toString('\n')
    )
  }
}
