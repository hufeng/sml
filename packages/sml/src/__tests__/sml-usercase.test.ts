import * as sml from '../index'
import { SmlUseCaseMeta } from '../lang/sml-usecase'
import { PumlUseCaseEmitter } from '../emitter/puml-usecase'

describe('sml usercase test suites', () => {
  const title = 'hello usercase diagram'

  it('test only usercases', () => {
    const usercase = sml.UsecaseDiagram(title, (ml) => {
      ml.usecase('blog', 'u1').usecase('music', 'u2').usecase('play', 'u3')
    })

    const meta = (usercase as any).meta as SmlUseCaseMeta
    expect(meta.usecases).toEqual([
      {
        name: 'u1',
        label: 'blog',
      },
      { name: 'u2', label: 'music' },
      { name: 'u3', label: 'play' },
    ])

    const code = new PumlUseCaseEmitter(usercase).emitCode()
    expect(code).toMatchSnapshot()
  })

  it('only actor', () => {
    const usercase = sml.UsecaseDiagram(title, (ml) => {
      ml.actor('first actor', 'user1')
        .actor('another actor', 'user2')
        .actor('Last actor', 'user3')
    })

    const meta = (usercase as any).meta as SmlUseCaseMeta
    expect(meta.actors).toEqual([
      {
        name: 'user1',
        label: 'first actor',
      },
      { name: 'user2', label: 'another actor' },
      { name: 'user3', label: 'Last actor' },
    ])

    const code = new PumlUseCaseEmitter(usercase).emitCode()
    expect(code).toMatchSnapshot()
  })

  it('test rect', () => {
    const usercase = sml.UsecaseDiagram(title, (ml) => {
      ml.actor('Food Critic', 'fc')
      ml.package('Restaurant')
        .usecase('Eat Food', 'uc1')
        .usecase('Pay for food', 'uc2')
        .usecase('Drink', 'uc3')

      ml.linkMany('fc', ['uc1', 'uc2', 'uc3'])
    })

    const meta = (usercase as any).meta as SmlUseCaseMeta
    expect(meta).toEqual({
      title,
      config: {
        actorStyle: 'default',
        packageStyle: 'Rectangle',
        direction: 'left->right',
      },
      usecases: [],
      notes: [],
      actors: [{ name: 'fc', label: 'Food Critic' }],
      packages: [
        {
          label: 'Restaurant',
          type: 'Rectangle',
          actors: [],
          usecases: [
            { name: 'uc1', label: 'Eat Food' },
            { name: 'uc2', label: 'Pay for food' },
            { name: 'uc3', label: 'Drink' },
          ],
        },
      ],
      links: [{ from: 'fc', to: ['uc1', 'uc2', 'uc3'] }],
    })
    const code = new PumlUseCaseEmitter(usercase).emitCode()
    expect(code).toMatchSnapshot()
  })

  it('test packages', () => {
    const usecase = sml.UsecaseDiagram(title, (ml) => {
      ml.actor('guest', 'g')
      ml.package(`Restaurant`, 'Rectangle')
        .usecase('Eat Food', 'uc1')
        .usecase('Pay for food', 'uc2')
      ml.package('Professional').actor('Chef', 'a')
      ml.linkMany('g', ['uc1', 'uc2', 'uc3']).link('a', 'uc1')
    })

    const code = new PumlUseCaseEmitter(usecase).emitCode()
    expect(code).toMatchSnapshot()
  })

  it('test actor style', () => {
    const usecase = sml.UsecaseDiagram('test actor style', (ml) => {
      ml.configuration.actorStyle('default')

      ml.actor('user', 'u1').usecase('Write Blog', 'c1').link('u1', 'c1')
    })
    const code = new PumlUseCaseEmitter(usecase).emitCode()
    expect(code).toMatchSnapshot()
  })

  it('test actor awesome style', () => {
    const usecase = sml.UsecaseDiagram('test actor asesome style', (ml) => {
      ml.configuration.actorStyle('awesome')
      ml.actor('User', 'a1').usecase('Learn JavaScript', `u1`)
      ml.link('a1', 'u1')
    })
    const code = new PumlUseCaseEmitter(usecase).emitCode()
    expect(code).toMatchSnapshot()
  })

  it('test note on simple actor or usecase', () => {
    const usecase = sml.UsecaseDiagram('test simple note', (ml) => {
      ml.actor('User', 'a1', ml.noteOf('a student user'))
      ml.usecase('Coding', 'u1', ml.noteOf('Coding Rust'))
    })
    const code = new PumlUseCaseEmitter(usecase).emitCode()
    expect(code).toMatchSnapshot()
  })
})
