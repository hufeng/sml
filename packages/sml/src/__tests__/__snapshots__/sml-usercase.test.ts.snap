// Vitest Snapshot v1

exports[`sml usercase test suites > only actor 1`] = `
"@startuml
left to right direction

actor :first actor: as user1
actor :another actor: as user2
actor :Last actor: as user3

@enduml"
`;

exports[`sml usercase test suites > test actor awesome style 1`] = `
"@startuml
skinparam actorStyle awesome
left to right direction

actor :User: as a1

usecase (Learn JavaScript) as u1

a1 --> u1
@enduml"
`;

exports[`sml usercase test suites > test actor style 1`] = `
"@startuml
left to right direction

actor :user: as u1

usecase (Write Blog) as c1

u1 --> c1
@enduml"
`;

exports[`sml usercase test suites > test note on simple actor or usecase 1`] = `
"@startuml
left to right direction

actor :User: as a1

usecase (Coding) as u1

note right of (a1)
  a student user
end note
note right of (u1)
  Coding Rust
end note

@enduml"
`;

exports[`sml usercase test suites > test only usercases 1`] = `
"@startuml
left to right direction

usecase (blog) as u1
usecase (music) as u2
usecase (play) as u3

@enduml"
`;

exports[`sml usercase test suites > test packages 1`] = `
"@startuml
left to right direction

actor :guest: as g

rectangle Restaurant {
  usecase \\"Eat Food\\" as uc1
  usecase \\"Pay for food\\" as uc2
}
rectangle Professional {
actor :Chef: as a
}

g --> uc1
g --> uc2
g --> uc3
a --> uc1
@enduml"
`;

exports[`sml usercase test suites > test rect 1`] = `
"@startuml
left to right direction

actor :Food Critic: as fc

rectangle Restaurant {
  usecase \\"Eat Food\\" as uc1
  usecase \\"Pay for food\\" as uc2
  usecase \\"Drink\\" as uc3
}

fc --> uc1
fc --> uc2
fc --> uc3
@enduml"
`;
